docs/src/appendix/packaging.md

```diff
@@ -0,0 +1,501 @@
+* One or more optional items that can be used with the package for development
+and debugging, the structure of which depends on the specific item:
+  * Debug info
+  * Documentation
+
+A package has a canonical directory layout, used both for packing and unpacking the contents
+of the package, as needed, see [Directory Layout](#directory-layout). However, the contents
+of the package can be directly extracted from the package file without extracting all of the
+contents. See [Binary Format](#binary-format) for details.
+
+## Top-Level Metadata
+
+* Schema Version
+  * The version of the package format itself, currently `1`.
+* Package Name (fully-qualified, e.g. `namespace/package`)
+* Package Version (using semver2 semantics)
+* Minimum required Miden VM version
+* Package Type (Executable vs Library)
+* Start function/entrypoint (optional for libraries)
+  * For program packages, specifies the entrypoint, i.e. `main`
+  * For library packages, if given, specifies a function that must be called to initialize the package before any of its exports may be called within a given context
+* Features (optional extras)
+  * Documentation
+  * Debug information
+* Dependencies (name, version, digest)
+
+## Type Descriptor
+
+The type descriptor file is the canonical, binary form of a WebAssembly Component Model [type definition](https://github.com/WebAssembly/component-model/blob/main/design/mvp/Explainer.md#type-definitions).
+It is derived from a [WebAssembly Interface Types](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md)
+file, maintained by hand by the package author.
+
+This descriptor is used for two purposes:
+
+1. To generate language-specific bindings to the interface(s) described. WIT is an Interface Description Language (IDL),
+describing interfaces in a language-agnostic way, but defined in terms of a [Canonical ABI](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md),
+allowing efficient cross-language interop. We follow the semantics of the Wasm canonical ABI, but layered on top of
+our canonical Miden ABI, which describes how the Wasm abstract machine is translated to the Miden abstract machine.
+
+2. To provide the compiler with the necessary type information needed to emit lifting/lowering code at call boundaries,
+as well as mapping imports (from the packages' dependencies) to their corresponding MAST roots, so that calls use those
+roots rather than the fully-qualified names. This ensures that the code that was compiled and published is always the
+exact code that gets executed, in perpetuity.
+
+Because the type descriptor is in binary form, to view it in the more human-friendly WIT format, you
+will need to use [wasm-tools](https://github.com/bytecodealliance/wasm-tools), like so:
+
+    wasm-tools component wit extracted-package/component.wasm
+
+The contents of `component.wasm` here is not an actual Wasm module, but simply the binary representation
```

Comment:
First, it's important to keep in mind that the package format must be:

* Compact
* Able to be created and consumed by hand, and by tooling in various languages (the full set of which is unknown, and which we do not control).
* Provide metadata sufficient to generate language-specific bindings to the interface exposed by that package, without the source code being available.

The binary format for WIT is a Wasm component type definition. There are multiple reasons why this form is preferred to the textual format:

* It is a more stable format (WIT will continue to evolve rapidly for some time, but the binary format is essentially final at this point)
* It is fully validated and processed, e.g. imports have been resolved. This means we can process the contents of a package without needing to do any of that work, which requires additional tooling.
* It is compact
* It is a richer format than WIT (things can be expressed in the binary format that cannot be in the textual format, at least at this point in time)
* It can be used to generate bindings without additional processing (to do so, the WIT must be processed into a form equivalent to what is represented in the binary format)
* The textual WIT can be recovered from the binary format if desired for readability (albeit not necessarily the exact same text, but an accurate representation nonetheless)

Raw textual WIT, in comparison, has basically none of those properties. A WIT "package" isn't really a thing, or more precisely, it is the specification for how WIT should be laid out on disk in a canonical form for processing into the binary format (which is what is [actually published to the registry](https://github.com/bytecodealliance/registry/blob/main/docs/README.md#webassembly-binary-packages). See also, [the `wit` cli](https://github.com/peterhuene/wit) which is a prototype tool for interacting with a Wasm component registry.

> we don't need anything for the dependencies besides MAST roots and invocation method whick we want to encode in the WIT package itself.

That information _can't_ be represented in WIT alone. We could _maybe_ encode the MAST roots in the binary format, I'd need to confirm that, but I'm not aware of any means by which we could do so in the textual format of WIT.

For the foreseeable future, interfaces described in WIT will always be `call`ed. It won't be possible to express things like the standard library as a package. Package dependencies will always be true components, i.e. with shared-nothing semantics, and relying on the canonical ABI. For the time being, things like the standard library will have to be handled as native dependencies (e.g. Rust crates).

Relaxing this restriction will require us to introduce some new functionality to the VM (and/or modify existing functionality). One hack/workaround we've discussed is allowing one to express functions which are WIT-compatible, in WIT, but then use a custom bindings generator to emit more bindings that operate under the assumption that the caller and callee are using the same shared memory (and which would be lowered as `exec` rather than `call`) - but at best that would have to be extra metadata in the package (probably with an associated feature flag), not something present in the WIT. The problem with that, however, is that it immediately breaks compatibility with existing tooling (for example, generating bindings from the WIT using a tool that is not aware of that custom logic would emit code for the caller that is incompatible with what is expected by the callee).

Until we can land on a solid longer term solution for this, and other related issues, we just simply can't support such packages. I would personally prefer to punt on it, until we revisit some details of how the VM manages contexts and call-like instructions, at which point we can provide ourselves with the primitives necessary to solve this properly, but it will be some time before we can do that. We don't have an urgent need to solve this problem right now anyway, so it isn't something I'm especially concerned about, but it will become more of an issue over time.

However, in addition to knowing MAST roots, we _do_ need other dependency information. We may be able to encode basic information in the component definition (since it permits us a way to express component dependencies with both semantic versions and digests), but it is limited in what can be expressed, and cannot express dependencies on things which are not components, nor can it tell us how to fetch those dependencies. To use it as an example again, the `wit` tool I linked earlier also demonstrates the necessity for this, as it also has a manifest to provide information about what packages to fetch and how, because WIT on its own doesn't provide all of the pieces.

In the format I've described so far, I've actually omitted the metadata for "how" to fetch a dependency (aside from vendored vs not-vendored dependencies), because we don't any any package management infrastructure or tooling defined yet. However, we will (long-term) need a way to express those details - just like Cargo does. There is a strong chance we will base this on what the Wasm component registry spec ends up looking like, but it's not a given. Either way, without that information, there will be no means by which to fetch a dependency graph from a given package (which may be located on disk, downloaded from an HTTP server, cloned from source control, or fetched from a proper registry). For that reason, some degree of dependency metadata is important to represent in the package manifest.

That said, if there is information being duplicated, and we can avoid doing so while keeping frequently-accessed metadata easily accessible, I would like to make sure we make any changes necessary to accomplish that goal.

Reply:
Thank you for such a detailed reply!

> First, it's important to keep in mind that the package format must be:
> 
> * Compact
> * Able to be created and consumed by hand, and by tooling in various languages (the full set of which is unknown, and which we do not control).
> * Provide metadata sufficient to generate language-specific bindings to the interface exposed by that package, without the source code being available.
> 
> The binary format for WIT is a Wasm component type definition. There are multiple reasons why this form is preferred to the textual format:
> 
> * It is a more stable format (WIT will continue to evolve rapidly for some time, but the binary format is essentially final at this point)
> * It is fully validated and processed, e.g. imports have been resolved. This means we can process the contents of a package without needing to do any of that work, which requires additional tooling.
> * It is compact
> * It is a richer format than WIT (things can be expressed in the binary format that cannot be in the textual format, at least at this point in time)
> * It can be used to generate bindings without additional processing (to do so, the WIT must be processed into a form equivalent to what is represented in the binary format)
> * The textual WIT can be recovered from the binary format if desired for readability (albeit not necessarily the exact same text, but an accurate representation nonetheless)
> 
> Raw textual WIT, in comparison, has basically none of those properties. A WIT "package" isn't really a thing, or more precisely, it is the specification for how WIT should be laid out on disk in a canonical form for processing into the binary format (which is what is [actually published to the registry](https://github.com/bytecodealliance/registry/blob/main/docs/README.md#webassembly-binary-packages). See also, [the `wit` cli](https://github.com/peterhuene/wit) which is a prototype tool for interacting with a Wasm component registry.

I agree. I'll dig into `wit` tool and warg packages. I suspect that I saw the WIT binary data and it is defined at the end of the Rust bindings source file, stored in the custom section of the Wasm component. 

> > we don't need anything for the dependencies besides MAST roots and invocation method whick we want to encode in the WIT package itself.
> 
> That information _can't_ be represented in WIT alone. We could _maybe_ encode the MAST roots in the binary format, I'd need to confirm that, but I'm not aware of any means by which we could do so in the textual format of WIT.

Yes. Althoug I'm recalling that we discussed that we might be able to encode it as an attributes in the comments but they would not be in the WIT binary anyway.

> For the foreseeable future, interfaces described in WIT will always be `call`ed. It won't be possible to express things like the standard library as a package. Package dependencies will always be true components, i.e. with shared-nothing semantics, and relying on the canonical ABI. For the time being, things like the standard library will have to be handled as native dependencies (e.g. Rust crates).

Sounds good. I'll remove the invocation method for imports/exports introduced in my first Wasm translation PR.